### ISSUE TITLE
integration of CTT tags detections

### ISSUE THREAD

**Issue by denislepage** [MotusDev/Motus-TO-DO#508]
As a first step to supporting CTT tags in sandbox, I have deactivated all of the recurring jobs in the sandbox database, including importing of Lotek detections, the data validation for the notification center, the computation of tracks, etc. They will only reactivated as one of the very last steps of the transition.

I will also shortly be clearing out the various sg data tables and associated summaries, so we can start switching all related procedures to the det tables. This means that data summaries and reports (including the R package API) will either return no data or very incomplete data based on the new tables as we do the transition.

There are about 80 million records available from 30,000 batches in the det tables so far, so we can start test indexing, etc. We will start bringing in progressively more, both from Lotek and CTT tags.

Note to self, I have safeguarded a backup of the motus sandbox database earlier today, in case we need to revert something. For Catherine, please confirm that the 2 database structures are now in sync (sandbox and live) before we start implementing structural changes.

**Comment by denislepage** [comment]
Progress update. 

Almost all of the sg tables have now been emptied, and I started an import of the data into the new det tables. At this point, I am not trying to resolve any of the deployment ID's, but most of the tables should be populated, with the main exception of the summaries (daily and runs).

The import process will likely take several days. We will likely need to interrupt it in order to work on the other modifications that we need to do throughout the database, and resume it whenever we are not.

We should also try to ensure that all the tables are being populated correctly sooner than later. Comparing them to the live database for instance for subsets of batchID's, etc.

**Comment by denislepage** [comment]
Progress update # 2

Import is still running. It is slow, but I am not sure how it compare against the previous import process. We're about 2/3 done. It should be easy to compare the import speed between old_import_log and det_import_log. So far, it's variable, with differences going both ways. The det import logs before 27 June can be ignored for that purpose, and the better comparison will be with the most recent batches from both processes.

We've renamed all the old sg_ tables to force anything still referring to them to break. There are a small number of tables that we kept under the name sg:

sg_api_fields
sg_uploads
sg_api_tokens

I have made several of the necessary changes in the motus master branch, wherever I could find references to the old sg table names, with the notable exception of org.bsc.motus.SGDataImport (I have started a new class for this called SGDataImport2).

**Comment by denislepage** [comment]
Also modified the following procs, which are called in sequence following a data import from SG:

1) prSGBatches_filters_ts : only flags lotek-related batches for duplicates, but for now, continues to exclude receivers located at Compudata.

- [ ] TODO: check whether version in batches field should be added to an index (wait until data import is done).

- [ ] TODO: test running speed. Shouldn't have to break the update call in small chunks for this.

2) prTaghitsDeployIDNew 

seems ok. 

- [ ] TODO: test running speed. Should be comparable to previous sg tables.

3) prSGBatchesActivityAll (and prSGBatchesActivityBatchID)

I limited the calculations of batch activity to sg data.

- [ ] TODO: check whether version in batches field should be added to an index (wait until data import is done).

- [ ] TODO: check what happens for ctt data when **det_batches_activity** is empty. Are these batches assumed to be OK, or are they filtered out due to inner joins? 

- [ ] TODO: check what happens for ctt data when **det_runs_filters** is empty. Are these runs assumed to be OK, or are they filtered out due to inner joins?  I think only filtered runs are listed in this table, so should be ok.

- [ ] TODO: test running speed. 

4) prTaghitsSummariesAll (and prTaghitsSummariesBatchID)

seems ok. 

- [ ] TODO: test running speed. Should be comparable to previous sg tables.


5) prDataValidation_pr 

I changed a couple references still using sg tables.

6) prSyncData_script

There are still several references to sg tables that need to be changed, but I haven't done so. Not sure if the sandbox or motus version is generally used.

7) tr_tags_update trigger in tags.

I changed the reference to det_batches_runs_tags

- [ ] TODO: also referred to sg_batches_tags, which didn't exist. I commented out for now.

I also fixed a number of objects that still had at least 1 occurrence (or more) of the sg tables, but this is not an exhaustive list. I looked through the database scripts, and I should have caught most if not all of them now:

 prMergeTags, prSelectProjectsTags, prDeleteTagsDeployment, prDeleteSensorDeployment, prAdminTableStats, prTaghitsSummariesSensorID, prTaghitsSummariesTagID, prTaghitsSummariesSensorDeployID, etc. and at least a few views.

**Comment by denislepage** [comment]
Another important issue pending:

The following tables have triggers that I put in place to deal with converting primary keys that are referenced in child tables when importing from SG. E.g. the hits table contains runID and batchID, but need to be converted from sgRunID and sgBatchID, which is handled by the triggers. 

The problem is that the same rules don't apply when importing data from CTT. Those triggers are referencing det_batches_sg_batch_id and det_runs_sg_run_id, which are not populated for CTT records. A couple options exist. The safest (but not necessarily most efficient) is to modify the triggers to do different steps for CTT and SG records. Another one is to disable and then re-enable triggers for CTT, but I believe this is a bad idea, and what to avoid this.

Perhaps a much better idea is to replace the triggers by procedures that call be called independently for SG and CTT inserts.

- [ ] det_pulse_counts (batchID)
- [ ] det_gps (batchID)
- [ ] det_taghits (batchID, runID)
- [ ] det_batches_runs (runID)
- [ ] det_runs (batchIDBegin)

**Comment by denislepage** [comment]
I have changed the java class to use procs to insert and upsert instead of relying on triggers. I have created new AFTER triggers and disabled the INSTEAD OF triggers (we can delete them once I have confirmed that it all works).

NOTE: We'll have to check that the update statements in the upsert for sg_runs that have completed until now were working as expected.

**Comment by denislepage** [comment]
Other issues needing attention.

- [ ] Deleting deprecated batches.

- [ ] Deleting spurious tags registered without project during processing.

- [ ] make a new table to track instances of multiple projects for a given tag, and incorporate this into the permission handling, etc.

**Comment by denislepage** [comment]
I checked a few stats on the import speed using the logs. the newer imports are on average just 0.1 second per batch faster than the previous ones, so really about the same. There's about an equal number of instances where the new imports were faster vs. slower, with the majority being under 1 second. One possible improvement will be to rely on the local IP address. There's a fix in place, but we'll need to restart the server to confirm.

**Comment by denislepage** [comment]
Quick update.

I have started the deploy ID assignments in sandbox. Based on progress so far, it should take another day+ to complete.

In parallel, I have also started the import in the DET tables of the live database, at least on a trial basis. So far so good, but I don't have an ETA yet. We won't be ready to do the deploy ID and summary procs until we have made several adjustments on the live db. To discuss.
